File: rust/src/registry.rs
use std::fs;
use std::path::Path;

use anyhow::{Context as _, Result};
use serde::{Deserialize, Serialize};

use crate::crypto;
use crate::store::ContextStore;

/// Default public registry URL
pub const DEFAULT_REGISTRY: &str = "https://openfuse-registry.wzmcghee.workers.dev";

/// Agent manifest — the "DNS record" for an agent in the registry.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Manifest {
    pub name: String,
    /// Where to reach this agent (http://, ssh://, gs://, s3://)
    pub endpoint: String,
    /// Ed25519 signing public key (hex)
    #[serde(rename = "publicKey")]
    pub public_key: String,
    /// age encryption public key (age1...)
    #[serde(rename = "encryptionKey", skip_serializing_if = "Option::is_none")]
    pub encryption_key: Option<String>,
    /// SHA-256 fingerprint of signing key
    pub fingerprint: String,
    pub created: String,
    pub capabilities: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Ed25519 signature over the canonical manifest (proves ownership)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signature: Option<String>,
    /// Timestamp used during signing (needed to verify)
    #[serde(rename = "signedAt", skip_serializing_if = "Option::is_none")]
    pub signed_at: Option<String>,
}

/// Resolve registry target from flag, env var, or default.
/// Returns either a URL (http(s)://) or a local path.
pub fn resolve_registry(flag: Option<&str>) -> String {
    if let Some(p) = flag {
        p.to_string()
    } else if let Ok(p) = std::env::var("OPENFUSE_REGISTRY") {
        p
    } else {
        DEFAULT_REGISTRY.to_string()
    }
}

fn is_http(registry: &str) -> bool {
    registry.starts_with("http://") || registry.starts_with("https://")
}

/// Build and sign a manifest from the current store config.
pub fn build_manifest(store: &ContextStore, endpoint: &str) -> Result<Manifest> {
    let config = store.read_config()?;
    let name = &config.name;
    let public_key = config
        .public_key
        .as_deref()
        .context("No signing key — run `openfuse init` first")?;
    let encryption_key = config.encryption_key.clone();
    let fingerprint = crypto::fingerprint(public_key);

    let mut manifest = Manifest {
        name: name.clone(),
        endpoint: endpoint.to_string(),
        public_key: public_key.to_string(),
        encryption_key,
        fingerprint,
        created: chrono::Utc::now().to_rfc3339(),
        capabilities: vec![
            "inbox".to_string(),
            "shared".to_string(),
            "knowledge".to_string(),
        ],
        description: None,
        signature: None,
        signed_at: None,
    };

    // Sign the manifest
    let canonical = canonical_manifest(&manifest);
    let signed = crypto::sign_message(store.root(), &manifest.name, &canonical)?;
    manifest.signature = Some(signed.signature);
    manifest.signed_at = Some(signed.timestamp);

    Ok(manifest)
}

/// Register this agent. Handles both local dir and HTTP registries.
pub async fn register(store: &ContextStore, endpoint: &str, registry: &str) -> Result<Manifest> {
    let manifest = build_manifest(store, endpoint)?;

    if is_http(registry) {
        register_http(&manifest, registry).await?;
    } else {
        register_local(&manifest, Path::new(registry))?;
    }

    Ok(manifest)
}

fn register_local(manifest: &Manifest, registry: &Path) -> Result<()> {
    let agent_dir = registry.join(&manifest.name);
    fs::create_dir_all(&agent_dir)?;
    let json = serde_json::to_string_pretty(manifest)?;
    fs::write(agent_dir.join("manifest.json"), format!("{}\n", json))?;
    Ok(())
}

async fn register_http(manifest: &Manifest, registry: &str) -> Result<()> {
    let client = reqwest::Client::new();
    let url = format!("{}/register", registry.trim_end_matches('/'));
    let body = serde_json::to_string(manifest)?;

    let resp = client
        .post(&url)
        .header("Content-Type", "application/json")
        .body(body)
        .send()
        .await
        .context("Failed to connect to registry")?;

    let status = resp.status();
    let text = resp.text().await.unwrap_or_default();

    if status.is_success() {
        Ok(())
    } else {
        // Try to extract error message from JSON
        let err_msg = serde_json::from_str::<serde_json::Value>(&text)
            .ok()
            .and_then(|v| v["error"].as_str().map(String::from))
            .unwrap_or(text);
        anyhow::bail!("Registry returned {}: {}", status, err_msg)
    }
}

/// Discover an agent by name. Handles both local dir and HTTP registries.
pub async fn discover(name: &str, registry: &str) -> Result<Manifest> {
    if is_http(registry) {
        discover_http(name, registry).await
    } else {
        discover_local(name, Path::new(registry))
    }
}

fn discover_local(name: &str, registry: &Path) -> Result<Manifest> {
    let manifest_path = registry.join(name).join("manifest.json");
    let raw = fs::read_to_string(&manifest_path)
        .with_context(|| format!("Agent '{}' not found in registry at {}", name, registry.display()))?;
    Ok(serde_json::from_str(&raw)?)
}

async fn discover_http(name: &str, registry: &str) -> Result<Manifest> {
    let client = reqwest::Client::new();
    let url = format!("{}/discover/{}", registry.trim_end_matches('/'), name);
    let resp = client.get(&url).send().await.context("Failed to connect to registry")?;

    let status = resp.status();
    let text = resp.text().await.unwrap_or_default();

    if status.is_success() {
        Ok(serde_json::from_str(&text)?)
    } else {
        let err_msg = serde_json::from_str::<serde_json::Value>(&text)
            .ok()
            .and_then(|v| v["error"].as_str().map(String::from))
            .unwrap_or(format!("HTTP {}", status));
        anyhow::bail!("{}", err_msg)
    }
}

/// List all agents in the registry.
pub async fn list_agents(registry: &str) -> Result<Vec<Manifest>> {
    if is_http(registry) {
        list_http(registry).await
    } else {
        list_local(Path::new(registry))
    }
}

fn list_local(registry: &Path) -> Result<Vec<Manifest>> {
    let mut agents = vec![];
    if !registry.exists() {
        return Ok(agents);
    }
    for entry in fs::read_dir(registry)? {
        let entry = entry?;
        if !entry.file_type()?.is_dir() {
            continue;
        }
        let manifest_path = entry.path().join("manifest.json");
        if manifest_path.exists() {
            if let Ok(raw) = fs::read_to_string(&manifest_path) {
                if let Ok(m) = serde_json::from_str::<Manifest>(&raw) {
                    agents.push(m);
                }
            }
        }
    }
    agents.sort_by(|a, b| a.name.cmp(&b.name));
    Ok(agents)
}

async fn list_http(registry: &str) -> Result<Vec<Manifest>> {
    let client = reqwest::Client::new();
    let url = format!("{}/list", registry.trim_end_matches('/'));
    let resp = client.get(&url).send().await?;
    let body: serde_json::Value = resp.json().await?;
    // The /list endpoint returns { agents: [...], count: N }
    // But agents are summaries, not full manifests. Return what we have.
    let mut agents = vec![];
    if let Some(arr) = body["agents"].as_array() {
        for a in arr {
            if let (Some(name), Some(endpoint), Some(fp)) = (
                a["name"].as_str(),
                a["endpoint"].as_str(),
                a["fingerprint"].as_str(),
            ) {
                agents.push(Manifest {
                    name: name.to_string(),
                    endpoint: endpoint.to_string(),
                    public_key: String::new(),
                    encryption_key: None,
                    fingerprint: fp.to_string(),
                    created: String::new(),
                    capabilities: vec![],
                    description: None,
                    signature: None,
                    signed_at: None,
                });
            }
        }
    }
    Ok(agents)
}

/// Verify a manifest's signature (proves the registrant owns the key).
pub fn verify_manifest(manifest: &Manifest) -> bool {
    let Some(ref sig) = manifest.signature else {
        return false;
    };
    let Some(ref signed_at) = manifest.signed_at else {
        return false;
    };
    let canonical = canonical_manifest(manifest);
    let signed = crypto::SignedMessage {
        from: manifest.name.clone(),
        timestamp: signed_at.clone(),
        message: canonical,
        signature: sig.clone(),
        public_key: manifest.public_key.clone(),
        encrypted: false,
    };
    crypto::verify_message(&signed)
}

/// Check if a newer version is available. Non-blocking, best-effort.
pub async fn check_update(current: &str) -> Option<String> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(2))
        .build()
        .ok()?;
    let resp = client.get(DEFAULT_REGISTRY).send().await.ok()?;
    let body: serde_json::Value = resp.json().await.ok()?;
    let latest = body["latest"].as_str()?;
    if latest != current {
        Some(latest.to_string())
    } else {
        None
    }
}

/// Canonical string representation of a manifest for signing.
fn canonical_manifest(m: &Manifest) -> String {
    format!(
        "{}|{}|{}|{}",
        m.name,
        m.endpoint,
        m.public_key,
        m.encryption_key.as_deref().unwrap_or("")
    )
}

File: src/registry.ts
// --- Registry: DNS + keyserver hybrid ---
// The registry solves agent discovery without requiring a DHT or blockchain.
// It's a signed directory: agents register name→endpoint+publicKey mappings,
// similar to DNS (name resolution) + PGP keyservers (key distribution).
// Crucially, imported keys are UNTRUSTED by default — the local agent must
// explicitly `openfuse key trust` after out-of-band verification (fingerprint check).
// This is TOFU (Trust On First Use) done right: the registry distributes keys,
// but never asserts trust. Trust is a local decision.

import { signMessage, fingerprint } from "./crypto.js";
import { ContextStore } from "./store.js";

// Registry API lives on .dev (product layer). DNS TXT records on .net (protocol layer).
export const DEFAULT_REGISTRY = "https://registry.openfused.dev";

export interface Manifest {
  name: string;
  endpoint: string;
  publicKey: string;
  encryptionKey?: string;
  fingerprint: string;
  created: string;
  capabilities: string[];
  description?: string;
  signature?: string;
  signedAt?: string;
  revoked?: boolean;
  revokedAt?: string;
  rotatedFrom?: string;
}

export function resolveRegistry(flag?: string): string {
  return flag || process.env.OPENFUSE_REGISTRY || DEFAULT_REGISTRY;
}

export async function register(store: ContextStore, endpoint: string, registry: string): Promise<Manifest> {
  const config = await store.readConfig();
  if (!config.publicKey) throw new Error("No signing key — run `openfuse init` first");

  const manifest: Manifest = {
    name: config.name,
    endpoint,
    publicKey: config.publicKey,
    encryptionKey: config.encryptionKey,
    fingerprint: fingerprint(config.publicKey),
    created: new Date().toISOString(),
    capabilities: ["inbox", "shared", "knowledge"],
  };

  // Canonical string prevents field-reordering attacks — pipe-delimited, deterministic order.
  // Signature proves the registrant owns the private key (anti-squatting).
  const canonical = `${manifest.name}|${manifest.endpoint}|${manifest.publicKey}|${manifest.encryptionKey || ""}`;
  const signed = await signMessage(store.root, manifest.name, canonical);
  manifest.signature = signed.signature;
  manifest.signedAt = signed.timestamp;

  const resp = await fetch(`${registry.replace(/\/$/, "")}/register`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(manifest),
  });

  if (!resp.ok) {
    const body = await resp.json().catch(() => ({ error: `HTTP ${resp.status}` })) as { error?: string };
    throw new Error(body.error || `Registry returned ${resp.status}`);
  }

  return manifest;
}

// Discovery: try DNS TXT first (decentralized, no registry needed), fall back to Worker API.
// DNS format: v=of1 e={endpoint} pk={pubkey} ek={agekey} fp={fingerprint}
// Self-hosted: _openfuse.{name}.{their-domain} — user manages their own TXT records.
// Our zone: _openfuse.{name}.openfused.net — managed by the registry Worker on registration.
export async function discover(name: string, registry: string): Promise<Manifest> {
  // If name contains a dot, it's a domain — try DNS TXT directly
  // Otherwise try DNS at openfused.net, then fall back to registry API
  const dnsNames = name.includes(".")
    ? [`_openfuse.${name}`]
    : [`_openfuse.${name}.openfused.net`];

  for (const dnsName of dnsNames) {
    const manifest = await discoverViaDns(dnsName, name);
    if (manifest) return manifest;
  }

  // Fall back to registry API
  const resp = await fetch(`${registry.replace(/\/$/, "")}/discover/${name}`);
  if (!resp.ok) {
    const body = await resp.json().catch(() => ({ error: `HTTP ${resp.status}` })) as { error?: string };
    throw new Error(body.error || `Agent '${name}' not found`);
  }
  return (await resp.json()) as Manifest;
}

async function discoverViaDns(dnsName: string, agentName: string): Promise<Manifest | null> {
  try {
    // Use DNS-over-HTTPS (Cloudflare 1.1.1.1) to resolve TXT records
    const resp = await fetch(`https://1.1.1.1/dns-query?name=${encodeURIComponent(dnsName)}&type=TXT`, {
      headers: { "Accept": "application/dns-json" },
    });
    if (!resp.ok) return null;
    const data = await resp.json() as { Answer?: { data: string }[] };
    if (!data.Answer || data.Answer.length === 0) return null;

    // Parse v=of1 format from TXT record
    const txt = data.Answer[0].data.replace(/"/g, "");
    if (!txt.startsWith("v=of1")) return null;

    const fields: Record<string, string> = {};
    for (const part of txt.split(" ")) {
      const [k, v] = part.split("=", 2);
      if (k && v) fields[k] = v;
    }

    if (!fields.e || !fields.pk) return null;

    return {
      name: agentName,
      endpoint: fields.e,
      publicKey: fields.pk,
      encryptionKey: fields.ek || undefined,
      fingerprint: fields.fp || "",
      created: "",
      capabilities: ["inbox", "shared", "knowledge"],
    };
  } catch {
    return null;
  }
}

// Revocation is permanent and self-authenticated: the agent signs its own revocation
// with the key being revoked. No admin needed — if you have the private key, you can kill it.
export async function revoke(store: ContextStore, registry: string): Promise<void> {
  const config = await store.readConfig();
  if (!config.publicKey) throw new Error("No signing key");

  const revokeMsg = `REVOKE:${config.publicKey}`;
  const signed = await signMessage(store.root, config.name, revokeMsg);

  const resp = await fetch(`${registry.replace(/\/$/, "")}/revoke`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      name: config.name,
      signature: signed.signature,
      signedAt: signed.timestamp,
    }),
  });

  if (!resp.ok) {
    const body = await resp.json().catch(() => ({ error: `HTTP ${resp.status}` })) as { error?: string };
    throw new Error(body.error || `Revocation failed`);
  }
}

// Non-blocking version check with 2s timeout — never delays the CLI for a slow network.
export async function checkUpdate(currentVersion: string): Promise<string | null> {
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 2000);
    const resp = await fetch(DEFAULT_REGISTRY, { signal: controller.signal });
    clearTimeout(timeout);
    if (!resp.ok) return null;
    const body = (await resp.json()) as { latest?: string };
    if (body.latest && body.latest !== currentVersion) return body.latest;
    return null;
  } catch {
    return null;
  }
}
